extra_data/sourcedata.py

```diff
@@ -87,3 +95,109 @@ def keys(self, inc_timestamps=True):
+                return self.files[0].index_group_names(self.source)
+            else:
+                return {key.partition('.')[0] for key in self.sel_keys}
+        else:
+            # CONTROL data has no key group.
+            return {''}
+
+    def _glob_keys(self, pattern: str) -> Optional[set]:
+        if self._is_control and not pattern.endswith(('.value', '*')):
+            pattern += '.value'
+
+        if pattern == '*':
+            # When the selection refers to all keys, make sure this
+            # is restricted to the current selection of keys for
+            # this source.
+            matched = self.sel_keys
+        elif glob_wildcards_re.search(pattern) is None:
+            # Selecting a single key (no wildcards in pattern)
+            # This check should be faster than scanning all keys:
+            matched = {pattern} if pattern in self else set()
+        else:
+            key_re = re.compile(fnmatch.translate(pattern))
+            matched = set(filter(key_re.match, self.keys()))
+
+        if matched == set():
+            raise PropertyNameError(pattern, self.source)
+
+        return matched
+
+    def select_keys(self, keys) -> 'SourceData':
+        if isinstance(keys, str):
+            keys = self._glob_keys(keys)
+        elif keys:
+            # If a specific set of keys is selected, make sure
+            # they are all valid, adding .value as needed for CONTROl keys.
+            normed_keys = set()
+            for key in keys:
+                if self._has_exact_key(key):
+                    normed_keys.add(key)
+                elif self._is_control and self._has_exact_key(key + '.value'):
+                    normed_keys.add(key + '.value')
+                else:
+                    raise PropertyNameError(key, self.source)
+                keys = normed_keys
+        else:
+            # Catches both an empty set and None.
+            # While the public API describes an empty set to
+            # refer to all keys, the internal API actually uses
+            # None for this case. This method is supposed to
+            # accept both cases in order to natively support
+            # passing a DataCollection as the selector. To keep
+            # the conditions below clearer, any non-True value
+            # is converted to None.
+            keys = None
+
+        if self.sel_keys is None:
+            # Current keys are unspecific - use the specified keys
+            new_keys = keys
+        elif keys is None:
+            # Current keys are specific but new selection is not - use current
+            new_keys = self.sel_keys
+        else:
+            # Both the new and current keys are specific: take the intersection.
+            # The check above should ensure this never results in an empty set,
+            # but
+            new_keys = self.sel_keys & keys
+            assert new_keys
+
+        res = copy(self)
+        res.sel_keys = new_keys
+        return res
```

Comment:
These 3 methods that return a new `SourceData` object all use this pattern of making a copy and modifying it. Similar methods in DataCollection & KeyData work by creating a new object and passing all the unchanged fields back to it. I'm not sure which one I like better.

Using `copy()` tends to be fewer lines of code, and if we add a new attribute for `SourceData` objects, we don't have to change every such method to copy it across. But creating a new object each time is more explicit, and you can reuse any type conversions or checks in the `__init__` method. :shrug: